### PR TITLE
H364: Target-magnitude top-decile WSS hotspot reweighting

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**H364: Target-magnitude top-decile WSS hotspot reweighting** — up-weight surface points in the top decile of `||target_wss||` (norm of the true WSS vector) to focus gradient mass on vortex-shedding regions (wheel arches, A-pillar, mirror wakes) where WSS magnitude is highest and rel_l2 error is driven.

**Why this is structurally different from all closed scalar-reweight axes:**
- H339 (uniform WSS channel reweight): same weight at every surface point
- H341 (wz-only channel reweight): per-channel not per-point
- H346 (per-vertex focal EMA): dynamic weight conditioned on *historical prediction error*, not target magnitude

H364 uses a **static, target-conditioned per-point weight**: points with `||target_wss||` in top decile get `w=γ`, others get `w=1`. This explicitly concentrates gradient mass on the *physically meaningful high-WSS zones* — the same vortex-shedding hotspots that determine whether WSS_z improves.

**Discriminator context**: H361 (direction-magnitude decomposed loss) confirmed that the bottleneck is representation, not loss geometry. H364 asks whether the bottleneck is specifically *gradient mass concentration* — the current per-channel MSE distributes gradient equally across all surface points, including low-WSS background. If high-WSS hotspots need more representation capacity, up-weighting them at training time concentrates capacity there.

**Expected outcome**: Either (a) WSS_z improves, revealing gradient allocation is a key bottleneck, or (b) null/regression, confirming the bottleneck is representation depth/architecture (consistent with H361's discriminator).

**Two arms to screen**: γ=3 (Arm A) and γ=5 (Arm B), each 3-epoch cosine-tail fine-tune from H336 EP13 checkpoint. Arm B conditional on Arm A val_raw passing the close rule.

### Environment
All runs: DDP 8 GPUs via `torchrun --nproc_per_node=8`

### Step 1 — Implement per-point target-magnitude weight in `train.py`

In `train.py`, locate the surface loss computation. Add a static per-point weight mask based on the norm of the true WSS vector target. Apply this weight to the per-point MSE terms for WSS channels only (do not weight SP or VP predictions).

**Concrete implementation:**

```python
# In the surface loss block (where you compute per-point WSS MSE):
# Compute target WSS norm across all 3 channels
target_wss = target_surface[..., wss_channel_indices]  # shape (B, N, 3) or (N, 3)
target_wss_norm = torch.norm(target_wss, dim=-1)  # shape (B, N) or (N,)

# Compute top-decile threshold (90th percentile of ||target_wss|| in the batch)
threshold = torch.quantile(target_wss_norm.reshape(-1), 0.90)

# Build per-point weight
hotspot_weight = torch.where(target_wss_norm >= threshold,
                              torch.full_like(target_wss_norm, wss_hotspot_gamma),
                              torch.ones_like(target_wss_norm))

# Apply to WSS per-point MSE: multiply each point's squared error by hotspot_weight
# e.g. loss_wss_per_point = (pred_wss - target_wss).pow(2)  # shape (B, N, 3) or (N, 3)
# weighted_loss_wss = (loss_wss_per_point * hotspot_weight.unsqueeze(-1)).mean()
```

Add a CLI flag `--wss_hotspot_gamma` (default 1.0 = disabled). Use `--wss_hotspot_gamma 3.0` for Arm A and `--wss_hotspot_gamma 5.0` for Arm B.

**Important**: Log `wss_hotspot_gamma` to W&B config. Do NOT touch `tau_z_loss_weight` (locked at 1.67). Do NOT touch data loaders or protected files.

### Step 2 — Smoke test (2 steps, 2 GPU ranks)

```bash
SENPAI_MAX_EPOCHS=14 torchrun --standalone --nproc_per_node=2 train.py \
  --epochs 14 --epochs_already_done 13 \
  --resume_from_wandb yw2a5dyl --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_hotspot_gamma 3.0 \
  --wandb_group h364-wss-hotspot-reweight \
  --wandb_name "edward/h364-smoke" \
  --max_steps 30
```

Confirm train/loss decreases normally (should look like H336 fine-tune, ~0.015-0.020 at step 30).

### Step 3 — Arm A: γ=3 (8 GPUs, 3-epoch cosine-tail fine-tune)

**Pre-committed close rule**: If EP14 val_primary/abupt_axis_mean_rel_l2_pct > 6.05%, close immediately (same rule as H361/H358 — prevents wasting eval time on clear regressions).

```bash
SENPAI_MAX_EPOCHS=16 SENPAI_TIMEOUT_MINUTES=540 \
torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 \
  --resume_from_wandb yw2a5dyl --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_hotspot_gamma 3.0 \
  --wandb_group h364-wss-hotspot-reweight \
  --wandb_name "edward/h364-armA-gamma3"
```

### Step 4 — Arm A Phase 2: val close-rule check

After EP16 val:
- If val_primary abupt > 5.98%: apply the EP16 close rule — report results and stop (skip Arm B, skip TTA)
- If val_primary abupt ≤ 5.98%: proceed to Arm B

If Phase 1 val passes close rules, launch TTA eval using **H342 recipe** (K=5 × Student-t ν=4 × 8-res × mirror × antithetic) — same as the current SOTA recipe. The TTA eval chain should mirror what alphonse did for H356 (3-checkpoint output-space average of EP13+EP14+EP15, then K=5 TTA).

### Step 5 — Arm B: γ=5 (conditional on Arm A passing close rule)

**Only run if Arm A val_primary EP16 ≤ 5.98%.**

```bash
SENPAI_MAX_EPOCHS=16 SENPAI_TIMEOUT_MINUTES=540 \
torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 \
  --resume_from_wandb yw2a5dyl --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_hotspot_gamma 5.0 \
  --wandb_group h364-wss-hotspot-reweight \
  --wandb_name "edward/h364-armB-gamma5"
```

## Baseline

Current SOTA: **H342 PR #1526** (3-cp output-average × H336 TTA recipe)
- **val_abupt_calibrated = 5.8962%** (gate: must beat this)
- **test_abupt_calibrated = 5.7357%** (gate: must beat this)
- AND-logic: both must improve for merge

Per-channel breakdown:
- val_WSS=6.7038%, val_WSS_z=9.0807%, val_SP=3.9064%, val_VP=3.4601%
- test_WSS=6.5963%, test_WSS_z=8.6122%, test_SP=3.5770%, test_VP=3.4218%

Source checkpoint for fine-tuning: **H336 EP13 W&B run `yw2a5dyl`, alias `epoch-13`**

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

Reproduce current SOTA:
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 \
  --resume_from_wandb yw2a5dyl --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wandb_group h342-3cp-output-avg
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: stale_wip flag is heartbeat-timing artifact. Verified Arm A gamma=3 is RUNNING healthy with 8 DDP ranks (z1tltc7h, he2op633, r392d5br, x4mlzrxc, m2agb72n, u1tkaxtl, 7on73pcc, bphdhfes), rt=41.5min, started 2026-06-02T05:32:35Z. ETA Phase 1 completion ~12:30Z. Continue training; will revisit when terminal SENPAI-RESULT posts.

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["z1tltc7h"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":6.5858},"test_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":6.5858}}

## Results — H364 Arm A (γ=3): EP14 close rule triggered

**Outcome:** Hypothesis FALSIFIED. EP14 val_primary = **6.586%**, well above the 6.05% pre-committed close threshold. Training stopped at EP15 progress mark; Arm B (γ=5) and TTA both skipped per the close rule.

### EP14 validation (γ=3, 1 epoch fine-tune from H336 EP13 = `yw2a5dyl`)

| Metric | Arm A EP14 | H342 SOTA (val gate) | Δ vs gate |
|---|---|---|---|
| **val_abupt** | **6.5858%** | **5.8962%** | **+0.69 pp ✗** |
| val_WSS | 7.2842% | 6.7038% | +0.58 pp |
| val_WSS_x | 6.3668% | — | — |
| val_WSS_y | 8.0481% | — | — |
| val_WSS_z | 9.7136% | 9.0807% | +0.63 pp |
| val_SP | 4.2890% | 3.9064% | +0.38 pp |
| val_VP | 4.5113% | 3.4601% | +1.05 pp |

For context vs the source checkpoint (H336 EP13, run `yw2a5dyl`):
- H336 EP13 val_abupt = 6.017% → Arm A EP14 = 6.586% (+0.57 pp regression after 1 cosine-tail fine-tune epoch with γ=3)

### Hotspot mask diagnostics (sanity)

The reweighting mask is correctly defined:
- `wss_hotspot/fraction = 0.1000` (top-decile exactly as designed)
- `wss_hotspot/threshold = 4.844` Pa-equivalent (in physical units)
- `wss_hotspot/mean_norm_in = 7.622` Pa (top-decile mean)
- `wss_hotspot/mean_norm_out = 1.247` Pa (bottom-90% mean)

Top-decile points carry **6.1×** the mean WSS magnitude of bottom-90%, confirming the mask captures the high-magnitude hotspots (wheel arches, A-pillar, mirror wakes) as intended.

### Command

```bash
SENPAI_MAX_EPOCHS=16 SENPAI_TIMEOUT_MINUTES=540 \
torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs-already-done 13 \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --lr 1e-4 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --optimizer lion --weight-decay 0.0005 --grad-clip-norm 0.5 \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 --model-mlp-ratio 4 \
  --use-qk-norm --use-surf-to-vol-xattn \
  --rff-num-features 16 --rff-init-sigmas 0.25,0.5,1.0,2.0,4.0 \
  --pos-encoding-mode string_separable --drop-path-max 0.1 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --mirror-augmentation --batch-size 4 \
  --train-surface-points 65536 --train-volume-points 65536 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --save-every-epoch \
  --wss-hotspot-gamma 3.0 \
  --wandb-group h364-wss-hotspot-reweight \
  --wandb-name "edward/h364-armA-gamma3"
```

- Peak memory: ~74 GiB / GPU (DDP 8× H100, no OOM)
- W&B run: `z1tltc7h` (rank-0) — group `h364-wss-hotspot-reweight`
- Training stopped during EP15 (40% through) at 2026-06-02T06:24Z after EP14 val triggered close rule

### What happened — honest analysis

The hypothesis was: **gradient mass concentration on top-decile WSS hotspots will improve WSS_z**. The result falsifies this on two fronts.

**1. WSS_z got worse, not better.** WSS_z degraded from 9.0807% (H342 SOTA) to 9.7136% (+0.63 pp). The reweighted points received 3× the gradient signal, yet the model failed to capitalise — consistent with H361's discriminator that the bottleneck is *representation capacity*, not gradient allocation. A static target-conditioned weight tells the model *where* to look but does not give it any additional ability to *learn* high-magnitude vortex-shedding structure.

**2. SP and VP regressed even though they were not reweighted.** Despite the design touching only WSS channels, val_SP rose 0.38 pp and val_VP rose 1.05 pp. Mechanism: applying γ=3 to 10% of the WSS spatial points increases the effective WSS loss mass by ~(0.9·1 + 0.1·3) = 1.20× per channel, *relative to* SP and VP. The Lion optimizer (already tuned to H342's balance) is sensitive to this rebalance — the implicit shift in (surface vs. volume) and (WSS vs. cp) gradient ratios was enough to harm pressure channels by >1 pp.

**3. The hotspot mask is doing what it says on the tin.** Diagnostics confirm exactly 10% of valid points cross the threshold and they have 6.1× the mean WSS magnitude. So the mechanism is implemented correctly; the *idea* failed.

**Bottom line:** Combined with H361's null/regression on direction-magnitude decomposition, H364 closes another scalar-reweight axis. The H336 fine-tune basin is brittle to any loss-mass rebalancing — both H361 (cosine-direction) and H364 (point-magnitude) regress all channels. This is now a strong cumulative signal that **further scalar/static reweighting will not move the SOTA**.

### Suggested follow-ups

These all build on the H361 + H364 lesson that *static reweighting fails*:

1. **Skip Arm B (γ=5).** Per the close rule. The 6.5858% at γ=3 is already 0.7 pp over SOTA; γ=5 would only amplify the loss-mass imbalance and almost certainly regress further (the SP/VP penalty scales with γ).

2. **Auxiliary representation head, not loss reweighting.** If the bottleneck is WSS_z representation, the right lever is *architecture / capacity / inductive bias*, not loss weights. Concretely: a small auxiliary WSS-vorticity prediction head (predict ∇×v on the surface from the same latent) as a regularizer to force the latent to encode the vortex-shedding structure. This is structurally different from H361/H364 — it adds a *new prediction target* derived from the WSS gradient field, rather than rebalancing the existing per-point loss.

3. **Frequency-domain WSS_z analysis (read-only audit).** Before more experiments, project the H336 EP14 WSS_z prediction error onto a surface Laplacian basis (or just look at PSD vs target spatial frequency). If the error is dominated by high spatial frequencies that the current bottleneck width (16 RFF features, hidden=512) cannot represent, *no loss reweighting will help* — only widening the frequency representation will. This is cheap to run on existing checkpoints.

4. **Investigate why H342 SOTA is so brittle to gradient rebalancing.** Both H361 and H364 perturbed the loss landscape by O(20%) in gradient mass and got >0.5 pp regressions. That is a striking sensitivity that should be characterized — it suggests the H336 fine-tune basin is narrow and locally optimal under the *current* loss weighting, and progress here requires either (a) widening the basin via a different optimizer/schedule combo or (b) leaving the basin via architecture change.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1557 because ADVISOR close — Finding P (16th axis): `target-magnitude-wss-hotspot-reweight-overshoot`.

**Result**: Arm A γ=3 EP14 val_abupt=6.586% vs close threshold 6.05% (+54bp over threshold, +69bp vs H342 gate). Hypothesis FALSIFIED.

**Cumulative scalar-reweighting axis CLOSED**:
- H338 SP-loss-reweight → null
- H339 WSS-loss-reweight → null
- H341 wz-only WSS reweight → null
- H346 focal EMA per-vertex WSS_z → null
- H361 direction-magnitude decomposed loss → null
- H364 target-magnitude hotspot → null (this PR)

**Strong cumulative finding**: H336 fine-tune basin is BRITTLE to gradient mass rebalancing. Any O(20%) gradient redistribution → >0.5pp regression across all channels (including unreweighted ones, per your VP regression analysis). The bottleneck is REPRESENTATION CAPACITY, not loss geometry.

**Excellent analysis** on the SP/VP collateral regression (1.20× effective WSS gradient mass shifted Lion away from H342-tuned balance) — this is the key insight to bank for future loss-design hypotheses. Your suggestions #2 (aux representation head) and #3 (frequency-domain WSS_z error audit) are exactly the right next moves; the latter we'll fold into the H368 design.

GPU recycling to H368 next..
